ENH: Use array indexing preparation routines for flatiter objects

lysnikolaou opened the pull request:
<!--         ----------------------------------------------------------------
                MAKE SURE YOUR PR GETS THE ATTENTION IT DESERVES!
                ----------------------------------------------------------------

*  FORMAT IT RIGHT:
      https://www.numpy.org/devdocs/dev/development_workflow.html#writing-the-commit-message

*  IF IT'S A NEW FEATURE OR API CHANGE, TEST THE WATERS:
      https://www.numpy.org/devdocs/dev/development_workflow.html#get-the-mailing-list-s-opinion

*  HIT ALL THE GUIDELINES:
      https://numpy.org/devdocs/dev/index.html#guidelines

*  WHAT TO DO IF WE HAVEN'T GOTTEN BACK TO YOU:
      https://www.numpy.org/devdocs/dev/development_workflow.html#getting-your-pr-reviewed
-->
- Use `prepare_index` in `iter_subscript` and `iter_ass_subscript`. This fixes various cases that were broken before:
  - `arr.flat[[True, True]]`
  - `arr.flat[[1.0, 1.0]]`
  - `arr.flat[()] = 0`
- Add more extensive tests for `flatiter` indexing operations

Closes #28314.

lysnikolaou committed:
[ENH] Use array indexing preparation routines for flatiter objects

lysnikolaou committed:
Fix assign subscript and add tests for it

lysnikolaou committed:
Fix regression test

lysnikolaou committed:
Remove unnecessary dims array

lysnikolaou committed:
Add release note

ngoldbaum commented:
This is a big refactor, so I think we'll need at least two experienced developers to go over the C code changes, so that might take a while. I'll try to do a pass focusing on the correctness of the C code soon. On a first, high-level pass this looks like mostly simplification and cleanup.

I think you should also try running the indexing benchmarks to see if there are any significant regressions in existing benchmarks. I think `bench_indexing` already captures several workflows that go through the changed low-level C code path.

It would also be nice to get new entries in the `FlatIterIndexing` benchmark for newly added functionality.

lysnikolaou committed:
Remove unnecessary branch from iter_subscript

lysnikolaou committed:
Add more benchmarks

lysnikolaou committed:
Merge branch 'main' into use-prepare-index-flatiter

lysnikolaou commented:
These are the results of running the (old & new) benchmarks:

```
[100.00%] ··· bench_indexing.FlatIterIndexing.time_flat_slice_index                                                                                            3.95±0.04ms
| Change   | Before [93898621] <main>   | After [cfcdabf0] <use-prepare-index-flatiter>   |     Ratio | Benchmark (Parameter)                                       |
|----------|----------------------------|-------------------------------------------------|-----------|-------------------------------------------------------------|
| +        | 87.0±0.4ns                 | 43.0±0.2ms                                      | 494276    | bench_indexing.FlatIterIndexing.time_flat_empty_tuple_index |
| +        | 115±3ns                    | 479±8ns                                         |      4.15 | bench_indexing.FlatIterIndexing.time_flat_bool_index_0d     |
| +        | 39.4±0.3ms                 | 42.8±0.3ms                                      |      1.09 | bench_indexing.FlatIterIndexing.time_flat_ellipsis_index    |
| +        | 3.95±0.04ms                | 4.29±0.07ms                                     |      1.09 | bench_indexing.FlatIterIndexing.time_flat_slice_index       |
```

It looks like having special cases for tuple, ellipses etc. (instead of going through `prepare_index`) did have an impact on performance. Should we try and keep those special cases in?

ngoldbaum commented:
> Should we try and keep those special cases in?

Probably

lysnikolaou committed:
Merge branch 'main' into use-prepare-index-flatiter

lysnikolaou committed:
Add special cases to fix performance regression

lysnikolaou commented:
> > Should we try and keep those special cases in?
> 
> Probably

I added a couple of special cases for an empty tuple and boolean indexes. This fixes the two worst performance regressions. I feel that the rest are acceptable, since this goes through a much more complex code path to make sure that everything is set up correctly.

ngoldbaum commented:
I added the 2.3 milestone to make sure we don't drop reviewing this before cutting the release.

charris commented:
> I added the 2.3 milestone

@ngoldbaum I am about to push this off to 2.4 unless you want to put it in very soon.

ngoldbaum commented:
I spoke with Lysandros and he said it's OK to push this off. We'll coordinate on getting this reviewed soon.

seberg commented on numpy/_core/src/multiarray/iterators.c:
```diff
@@ -857,201 +811,154 @@ iter_ass_sub_int(PyArrayIterObject *self, PyArrayObject *ind,
+    PyArray_ITER_RESET(self);
+
+    if (index_type == HAS_INTEGER) {
+        if (check_and_adjust_index(&indices[0].value, self->size, -1, NULL) < 0) {
```
Here we probably need a cast again (or just a test, if there is a test for e.g. `int16` input and it passes, then that would be fine).

seberg commented on numpy/_core/src/multiarray/iterators.c:
```diff
@@ -857,201 +811,154 @@ iter_ass_sub_int(PyArrayIterObject *self, PyArrayObject *ind,
+        if (ret < 0) {
             PyErr_SetString(PyExc_ValueError,
-                            "Error setting single item of array.");
+                            "Error setting single item of flat iterator.");
```
Weird, but maybe let's keep it.  Feels like we should either chain the original array here, or jus tnot bother with setting a new error to begin with.

seberg commented on numpy/_core/src/multiarray/iterators.c:
```diff
@@ -857,201 +811,154 @@ iter_ass_sub_int(PyArrayIterObject *self, PyArrayObject *ind,
-    Py_INCREF(type);
-    arrval = (PyArrayObject *)PyArray_FromAny(val, type, 0, 0,
+    Py_INCREF(dtype);
+    PyArrayObject *arrval = (PyArrayObject *)PyArray_FromAny(val, dtype, 0, 0,
```
Would be good to pass the correct maxdims here, IIRC (fixes corner cases around object arrays, even if the choice of how that behaves is a matter of taste).

lysnikolaou replied:
How would we do that? `array_prepare_index` uses this exact construct.

seberg replied:
The `0, 0` would be `0, maxdims`, but matching the normal indexing makes sense indeed.

seberg commented on numpy/_core/src/multiarray/iterators.c:
```diff
@@ -561,195 +561,149 @@ iter_subscript_int(PyArrayIterObject *self, PyArrayObject *ind,
 
-    /* Check for Integer or Slice */
-    if (PyLong_Check(ind) || PySlice_Check(ind)) {
-        start = parse_index_entry(ind, &step_size, &n_steps,
```
Seems like `parse_index_entry` should be deleted, but is not yet.

seberg commented on numpy/_core/tests/test_indexing.py:
```diff
@@ -1434,3 +1418,190 @@ def test_setitem(self):
+        indices_2d = np.array([[1, 2], [3, 4]])
+        assert_array_equal(a.flat[indices_2d], indices_2d)
+
+        assert_array_equal(a.flat[[True, 1]], a.flat[[1, 1]])
```
As mentioned above, would be good to have a basic test here (or it's own test) for e.g. `int16` dtype inputs.

(And yes, you can force-cast to integer.)

seberg commented on numpy/_core/tests/test_indexing.py:
```diff
@@ -1434,3 +1418,190 @@ def test_setitem(self):
+    def test_flatiter_indexing_boolean(self):
+        a = np.arange(9).reshape((3, 3))
+        a.flat[True] = 10
+        assert_array_equal(a, np.array([[10, 1, 2], [3, 4, 5], [6, 7, 8]]))
```
Technically wrong. `a.flat[True]` should return the `a.reshape(1, a.size)` effectively.  So if anything it would assign it to everything.

In practice, I may be tempted to just deprecate it, since it seems somewhat useless?

seberg commented on numpy/_core/src/multiarray/mapping.c:
```diff
@@ -740,6 +704,39 @@ prepare_index(PyArrayObject *self, PyObject *index,
+ * @param allow_boolean whether to allow the boolean special case
+ *
+ * @returns the index_type or -1 on failure and fills the number of indices.
+ */
```
Maybe we should adjust that slightly and leave it on `prepare_index_noarray`?  It's obvious to look for docs there if you look at `prepare_index`, but not vice-versa?
(but just nitpicking/suggestion.)

seberg reviewed (commented):
Sorry for not looking at it much.  Overall looks nice, I need to do a pass to see for refcount issues, etc.

I am slightly worried that some of the bad bool cases should maybe have a `FutureWarning` (or just go to an error for a bit?!), to enforce correct behavior.

Overall, I am happy that this seemed to have worked well to integrate, the diff is a bit unwieldy, but it can't be helped.

lysnikolaou committed:
Merge branch 'main' into use-prepare-index-flatiter

lysnikolaou commented on numpy/_core/src/multiarray/iterators.c:
```diff
@@ -857,201 +811,154 @@ iter_ass_sub_int(PyArrayIterObject *self, PyArrayObject *ind,
+    else if (indices[0].type == HAS_0D_BOOL) {
+        ret = 0;
+        if (indices[0].value) {
+            ret = PyArray_Pack(PyArray_DESCR(self->ao), self->dataptr, val);
```
Because of performance again. I was using this exact branch there, but that introduced a performance regression. I'm not sure how to best handle that. Adding these "custom" branches is seemingly a bit random, but they affect performance quite a bit. Maybe the way to go is to leave those branches in and add a comment?

seberg replied:
There is an interesting question whether the main code should have a fast-path for these, but yeah, I am ok to keep the fast-path, especally if you have a meanignful case where it helps.

seberg commented on numpy/_core/src/multiarray/iterators.c:
```diff
@@ -561,195 +561,149 @@ iter_subscript_int(PyArrayIterObject *self, PyArrayObject *ind,
-        return obj;
+    if (PyTuple_Check(ind) && PyTuple_GET_SIZE(ind) == 0) {
+        Py_INCREF(self->ao);
+        return (PyObject *)self->ao;
```
Makes sense, so if you prefer to keep it that seems fine.  I think I would still lean towards just not doing this, because I don't really see where anyone would use `arr.flat[()]`.

lysnikolaou committed:
Address some feedback

- Add test with int16 index
- Cast integer index array to intp before using it
- Remove unused functions
- Change error message back to array
- Move docstring to prepare_index_noarray

lysnikolaou committed:
Address more feedback

lysnikolaou committed:
Address more feedback; raise error on non-array index

lysnikolaou committed:
Fix linting errors

lysnikolaou committed:
Update changelog item

lysnikolaou commented:
@seberg Sorry for taking so long with this and thanks a lot for the thorough review. It was really helpful!

I've addressed all of the feedback so it should be in a better state now. Looking foward to hearing your thoughts.

lysnikolaou committed:
Fix tests

seberg commented on doc/release/upcoming_changes/28590.improvement.rst:
```diff
@@ -0,0 +1,26 @@
+Improve `flatiter` indexing by using ndarray index preparation
```
Very opinionated, but I would prefer to home in on user-facing changes only.

seberg commented on doc/release/upcoming_changes/28590.improvement.rst:
```diff
@@ -0,0 +1,26 @@
+  Complicated non-array indices like ``arr.flat[[True, True]]`` and
+  ``arr.flat[[1.0, 1.0]]`` were incorrectly
+  treated as ``arr.flat[[1, 1]]``. They now raise an `IndexError`. Array
+  indices should be used instead.
```
Should point out that the boolean version may be re-enabled in the future.

seberg commented on numpy/_core/src/multiarray/iterators.c:
```diff
@@ -561,195 +489,157 @@ iter_subscript_int(PyArrayIterObject *self, PyArrayObject *ind,
+
+    // Single boolean index
+    else if (indices[0].type == HAS_0D_BOOL) {
+        DEPRECATE("Indexing flat iterators with a 0-dimensional boolean index is "
```
Needs:
* A comment `/* Deprecated 2025-..., NumPy 2.4 */`
* an error check
* And a test in `test_deprecations()` which would have failed the missing error check.
* Please end the message with `(Deprecated NumPy 2.4)` or similar, to give people a sense of how recent this happened.

seberg commented on numpy/_core/src/multiarray/iterators.c:
```diff
@@ -561,195 +489,157 @@ iter_subscript_int(PyArrayIterObject *self, PyArrayObject *ind,
                                  dtype,
                                  1, &n_steps,
                                  NULL, NULL,
                                  0, (PyObject *)self->ao);
```
Not your problem, so ignore.  I think there might be a bit of fuzziness around the fact that sometimes `dtypes` get's replaced on array creation.

seberg commented on numpy/_core/src/multiarray/iterators.c:
```diff
@@ -561,195 +489,157 @@ iter_subscript_int(PyArrayIterObject *self, PyArrayObject *ind,
+        PyErr_SetString(PyExc_IndexError,
+            "Non-array indices are not supported because they can lead to unexpected behavior. "
+            "This is expected to change in future versions.");
         goto finish;
```
This is tedious :/.  I don't think it is right here, we should move things down because `arr.flat[[1, 2, 3, 4]]` is really OK.

So there are two things here:
1. `a.flat[[1.]]` floats are at this point, rejected by the original preparation.  Changing that this stops working is also the biggest break (of sensible behavior).  I could see adding a branch to the index-prep code to deprecate it there (or we gamble and see if someone notices, even considering that array indexing is strict for a decade, I am not sure).
2. `a.flat[[True, False]]` was always nonsense, and I am certainly happy to not worry about deprecation. Two possible solutions:
   * Just move this (similarly) into the `HAS_BOOL` branch. That works, although the current will also fail for `arr.flat[np.array([True, False]),]` as it is a tuple.  Admittedly, that is rather niche.
   * We also push this into the prep-index code, where we have the original object at hand to give the warning.

lysnikolaou replied:
The float case is now fixed in that it works but is derecated (added a branch in the index-prep code). I'm a bit confused about the bool case though. When bool indices are used that do not match the iterator's shape, that raises an IndexError. That's what we want, so I think we're okay like this?

lysnikolaou committed:
Fix typing tests

seberg reviewed (commented):
Sorry this is so tedious.  I have to look further, but wanted to comment for those things.
I do wonder more now if we shouldn't push the errors/warning down into the index prep helper.  As annoying as it is, we just duplicate things and while that code is spaghetti, it is at least pretty linear spaghetti.

lysnikolaou committed:
Address review feedback

lysnikolaou committed:
Fix linter errors

seberg reviewed (commented):
Thanks, had a look through.  I think two of the comments are probably relevant (avoid the often unnecessary copy and actually raise the error we said we would raise).

But otherwise, I think this is turning out very nice and much simpler than the old code (even if that means making the index prep function slightly larger).

lysnikolaou committed:
Address feedback

seberg commented on numpy/_core/src/multiarray/iterators.c:
```diff
@@ -881,13 +884,16 @@ iter_ass_subscript(PyArrayIterObject *self, PyObject *ind, PyObject *val)
     }
 
     if (index_type == HAS_BOOL) {
+        // This cannot be reached, because boolean indices temporarily raise an error.
+        // TODO: Enable valid boolean indices
```
```suggestion
```
Also above, can be reached if an array is passed of course.  But can just apply that, and it might be th eonly thing.

lysnikolaou replied:
Whoops, you're right!

lysnikolaou committed:
Remove wrong comments

lysnikolaou committed:
Fix linting errors

lysnikolaou committed:
Fix test_deprecations

seberg commented on numpy/_core/src/multiarray/iterators.c:
```diff
@@ -561,195 +489,154 @@ iter_subscript_int(PyArrayIterObject *self, PyArrayObject *ind,
-            PyErr_SetString(PyExc_IndexError,
-                            "cannot use Ellipsis or newaxes here");
-            goto fail;
+    if (index_type & HAS_SLICE) {
```
```suggestion
    if (index_type == HAS_SLICE) {
```
But it doesn't matter, because it cannot have two kinds of indices.

seberg commented on numpy/_core/src/multiarray/iterators.c:
```diff
@@ -561,195 +489,154 @@ iter_subscript_int(PyArrayIterObject *self, PyArrayObject *ind,
-    Py_DECREF(obj);
+    PyErr_SetString(PyExc_IndexError,
+        "only integers, slices (`:`), ellipsis (`...`) and integer or boolean "
+        "arrays are valid indices"
```
Ah, I suspect this can't be reached, but it seems good to have as a back-stop.  I'll just apply the `==` above and hope it works fine.

(Just to be a bit pedantic.)

seberg commented on numpy/_core/src/multiarray/iterators.c:
```diff
@@ -857,201 +744,170 @@ iter_ass_sub_int(PyArrayIterObject *self, PyArrayObject *ind,
-        if (n_steps == ELLIPSIS_INDEX || n_steps == NEWAXIS_INDEX) {
-            PyErr_SetString(PyExc_IndexError,
-                            "cannot use Ellipsis or newaxes here");
+    if (index_type & HAS_SLICE) {
```
```suggestion
    if (index_type == HAS_SLICE) {
```
let's change it here too.

seberg commented on numpy/_core/src/multiarray/mapping.c:
```diff
@@ -418,17 +403,54 @@ prepare_index(PyArrayObject *self, PyObject *index,
+                // indexing routine would have succeeded as well), we need to issue a
+                // deprecation warning.
+                if (arr
+                    && is_flatiter_object
```
```suggestion                 
                    && is_flatiter_object
                    && PyArray_SIZE(tmp_arr) != 0
```
I guess just `[]` would otherwise fail, I am surprised there is no test for this.

seberg commented on numpy/_core/tests/test_indexing.py:
```diff
@@ -1453,3 +1437,229 @@ def test_setitem(self):
+        assert_array_equal(a.flat[indices_2d], indices_2d)
+
+        assert_array_equal(a.flat[[True, 1]], np.array([1, 1]))
+
```
```suggestion

        assert_array_equal(a.flat[[]], np.array([], dtype=a.dtype))

```

seberg reviewed (approved):
Thanks, there is one issue around `[]` I think, but I'll just apply the fix and hope CI passes :).

seberg committed:
Apply suggestions from code review

lysnikolaou commented:
That's probably because of this new `if` in `mapping.c`.

```
            if (PyArray_SIZE(tmp_arr) == 0
                || (is_flatiter_object && !PyArray_ISINTEGER(tmp_arr) && !PyArray_ISBOOL(tmp_arr))) {
```

The string array is cast to an integer array and since it's got no elements, there's no issues there.

seberg commented:
Oh, nvm, let's just just remove the test, or only test for a string array.  The array case just lets this one pass, because it can't doesn't distinguish between a `[]` and `array([]).flat`.

seberg commented:
The point is, that it isn't a string array, its a string flatiter... And we never handle it correctly, so I don't think we should care for this PR.

seberg committed:
Update numpy/_core/tests/test_indexing.py

seberg commented on numpy/_core/tests/test_indexing.py:
```diff
@@ -1453,3 +1437,231 @@ def test_setitem(self):
+        with pytest.warns(DeprecationWarning,
+                          match="Invalid non-array indices for iterator objects are "
+                                "deprecated"):
+            assert_equal(a.flat[b.flat], np.array([]))
```
@lysnikolaou I think this is my solution. Although I guess someone else might xfail the test also ;).

lysnikolaou commented on numpy/_core/src/multiarray/mapping.c:
```diff
@@ -418,17 +403,55 @@ prepare_index(PyArrayObject *self, PyObject *index,
+                            "only integers, slices (`:`), ellipsis (`...`)%s and integer or boolean "
+                            "arrays are valid indices",
+                            is_flatiter_object ? "" : ", numpy.newaxis (`None`)"
+                        );
```
It doesn't for this specific case. Do you know of any examples that would reach this?

```python
>>> import numpy as np
>>> a = np.arange(3)
>>> b = np.array(["a"], dtype="S")
>>> a[b]
Traceback (most recent call last):
  File "<python-input-8>", line 1, in <module>
    a[b]
    ~^^^
IndexError: arrays used as indices must be of integer (or boolean) type
>>> a.flat[b.flat]
Traceback (most recent call last):
  File "<python-input-9>", line 1, in <module>
    a.flat[b.flat]
    ~~~~~~^^^^^^^^
IndexError: only integers, slices (`:`), ellipsis (`...`) and integer or boolean arrays are valid indices
```

seberg replied:
Oh, I suppose it is practically impossible to reach indeed, since it only enters for length zero.
